src/Connection.php

```diff
@@ -0,0 +1,176 @@
+    private $port;
+    private $connect_timeout;
+
+    private $conn;
+
+    /**
+     * @param $url string postgres://user:pass@localhost:5432/dbname?param=val
+     */
+    public function __construct(string $url, int $connect_timeout = null)
+    {
+        $this->url = $url;
+        $this->connect_timeout = $connect_timeout;
+    }
+
+    public function connect()
+    {
+        if ($this->conn) {
+            return;
+        }
+
+        $host = parse_url($this->url, PHP_URL_HOST);
+        $port = parse_url($this->url, PHP_URL_PORT) ?: 5432;
+        $ip = gethostbyname($host);
+        $socket = "tcp://{$ip}:{$port}";
+        $this->conn = stream_socket_client(
+            $socket,
+            $error_code,
+            $error_message,
+            $this->connect_timeout ?: ini_get('default_socket_timeout')
+        );
+
+        if (!$this->conn) {
+            throw new PostgresException(
+                sprintf("Failed to connect to %s (%s)", $socket,
+                    $error_message)
+            );
+        }
+
+        // Write startup message
+        $user = parse_url($this->url, PHP_URL_USER);
+        $database = substr(parse_url($this->url, PHP_URL_PATH), 1);
+        parse_str(parse_url($this->url, PHP_URL_QUERY), $params);
+
+        $pk_format = 'nnZ*Z*Z*Z*';
+        $pk_args = [3, 0, 'user', $user, 'database', $database];
+        foreach ($params as $param => $val) {
+            $pk_format .= 'Z*Z*';
+            $pk_args[] = $param;
+            $pk_args[] = $val;
+        }
+        $pk_format .= 'x';
+
+        $msg = call_user_func_array('pack', array_merge([$pk_format], $pk_args));
+        $msg = pack('N', strlen($msg) + 4) . $msg;
+        $this->write($msg);
+
+        while (true) {
+            list($ident, , $buf) = $this->readMessage();
+            if ($ident === 'Z' || $ident === 'E') {
+                return;
+            }
+        }
+    }
+
+    public function query(string $sql): array
+    {
+        $msg = pack('Z*', $sql);
+        $msg = pack('a1N', 'Q', strlen($msg) + 4) . $msg;
+        $this->write($msg);
+
+        $rows = [];
+        while (true) {
+            list($ident, , $buf) = $this->readMessage();
+
+            if ($ident === 'T') {
+                $fields = [];
+                $num_fields = $buf->readInt16();
+                for ($i = 0; $i < $num_fields; $i++) {
+                    $fields[$i] = $buf->readString();
+                    $buf->readInt32();
+                    $buf->readInt16();
+                    $buf->readInt32();
+                    $buf->readInt16();
+                    $buf->readInt32();
+                    $buf->readInt16();
+                }
+            }
+
+            if ($ident === 'D') {
+                $row = [];
+                $num_cols = $buf->readInt16();
+                for ($i = 0; $i < $num_cols; $i++) {
+                    $col_val_len = $buf->readInt32();
+                    if ($col_val_len === 4294967295) {
```

Comment:
You might consider `0xFFFFFFFF` here instead of `4294967295`.

Reply:
lol yea, is clearer